BLE: Add KVStore Security DB

paul-szczepanek-arm opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->
This adds a new Security DB type for BLE Security Manager. This is optional and disabled by default. The presence of filesystem and kvstore DBs is now set in the mbed_lib.json file with the default values retaining the current setup.

Requires preceding PR: https://github.com/ARMmbed/mbed-os/pull/13037
(so ignore the first commit)
<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->
none
<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->
@pan- 
<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

pan- commented on features/FEATURE_BLE/source/generic/FileSecurityDb.cpp:
```diff
@@ -265,6 +267,29 @@ void FileSecurityDb::set_entry_peer_sign_counter(
+        sign_count_t sign_counter
+) {
+    this->SecurityDb::set_local_sign_counter(sign_counter);
+    db_write(&_local_sign_counter, DB_OFFSET_LOCAL_SIGN_COUNT);
```
We talked about it, is it something we want to do everytime the sign counter change ? 
I would also either override it in the KVStore implementation and just call the parent method or leave a note here explaining why it isn't overridden (too many writes...)

paul-szczepanek-arm replied:
that is fixed on the other branch, I'll rebase this

pan- reviewed (approved):
LGTM. Thanks for adding the changes related to the local identity and adding connection address to the whitelist.

0xc0170 reviewed (dismissed):
Can you extend commit 67ee6f6d1618d20cd89cbd33fc999b77906c69e0 with better commit message?

paul-szczepanek-arm committed:
Make kvstore and fs store configured in the lib json

paul-szczepanek-arm committed:
Relax the integrity requirement to allow private functions to call themselves

paul-szczepanek-arm committed:
Add a new implementation of security db for KVStore

This works similar to filesystem db but uses the KVStore which it assumes is initialised. This is checked by open_db. On initialisation it either reads the present db or writes a new db into all entries thus guaranteeing that after the initialisation we will not run out of space for the keys and no extra error handling is needed.

paul-szczepanek-arm committed:
Allow for security db type selection

The selection is now based on the lib json (which allows you to disable filesystem db or kvstore db) and the call to SecurityManager::init. It will always fall back on memory db if no other db is available.

paul-szczepanek-arm committed:
add SPDX identifier
